benchmarks/auth/go/server.go

```diff
@@ -0,0 +1,146 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"syscall"
+
+	pb "github.com/ease-lab/vSwarm/benchmarks/auth/proto"
+
+	tracing "github.com/ease-lab/vSwarm/utils/tracing/go"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+const (
+	default_port = "50051"
+)
+
+type Statement struct {
+	Action   string
+	Effect   string
+	Resource string
+}
+type policyDocument struct {
+	Version   string
+	Statement []Statement
+}
+type Context struct {
+	stringKey  string
+	numberKey  int
+	booleanKey bool
+}
+type authResponse struct {
+	principalId    string
+	policyDocument policyDocument
+	context        Context
+}
+
+func generatePolicy(principalId, effect, resource string) authResponse {
+	var authResponse authResponse
+
+	authResponse.principalId = principalId
+	if effect != "" && resource != "" {
+		var policyDocument policyDocument
+		policyDocument.Version = "2012-10-17"
+		// policyDocument.Statement = []
+		var statementOne Statement
+		statementOne.Action = "execute-api:Invoke"
+		statementOne.Effect = effect
+		statementOne.Resource = resource
+		policyDocument.Statement = []Statement{statementOne}
+		authResponse.policyDocument = policyDocument
+	}
+
+	// Optional output with custom properties of the String, Number or Boolean type.
+	authResponse.context = Context{stringKey: "stringval", numberKey: 123, booleanKey: true}
+	return authResponse
+}
+
+// server is used to implement auth.GreeterServer.
+type server struct {
+	pb.UnimplementedGreeterServer
+}
+
+// SayHello implements auth.GreeterServer
+func (s *server) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
+	gid := syscall.Getgid()
+	token := in.GetName()
+	fakeMethodArn := "arn:aws:execute-api:{regionId}:{accountId}:{apiId}/{stage}/{httpVerb}/[{resource}/[{child-resources}]]"
+	var msg string
+	var ret authResponse
+
+	switch token {
```

Comment:
Please implement the example functionality from AWS.
So switch on `allow` and `deny` tokens instead of `.f1` and then we should return the generated policy as string if possible. Can you check if that works?

```
// A simple token-based authorizer example to demonstrate how to use an authorization token 
// to allow or deny a request. In this example, the caller named 'user' is allowed to invoke 
// a request if the client-supplied token value is 'allow'. The caller is not allowed to invoke 
// the request if the token value is 'deny'. If the token value is 'unauthorized' or an empty
// string, the authorizer function returns an HTTP 401 status code. For any other token value, 
// the authorizer returns an HTTP 500 status code. 
// Note that token values are case-sensitive.

exports.handler =  function(event, context, callback) {
    var token = event.authorizationToken;
    switch (token) {
        case 'allow':
            callback(null, generatePolicy('user', 'Allow', event.methodArn));
            break;
        case 'deny':
            callback(null, generatePolicy('user', 'Deny', event.methodArn));
            break;
        case 'unauthorized':
            callback("Unauthorized");   // Return a 401 Unauthorized response
            break;
        default:
            callback("Error: Invalid token"); // Return a 500 Invalid token response
    }
};
```